### ISSUE TITLE
Unresponsive TrackPad

### ISSUE THREAD

**Issue by precondition** [david-janssen/kmonad#130]
I believe there is something wrong with my permissions which cause a few problems, the most annoying of which is the unresponsiveness of the trackpad when I want to enable KMonad on my laptop keyboard.

To preface, I am on Linux Xubuntu.

First of all, I did read the FAQ section on Linux permissions. I added my user to the `uinput` group. I also created a new file at `/lib/udev/rules.d/50-kmonad-rules.rules` and copy-pasted in the rules but that didn't seem to have an effect so I tried appending the rules  written in the FAQ to the `/lib/udev/rules.d/50-udev-default.rules` but that seems to have been in vain. I did also run `sudo modprobe uinput` but I didn't notice any effect from that.

My workaround has thus been to run the following command to give my user `rw` permissions to my laptop keyboard:
```sh
sudo chgrp uinput /dev/input/by-path/platform-i8042-serio-0-event-kbd && sudo chmod g+rw /dev/input/by-path/platform-i8042-serio-0-event-kbd
```

Once that's run, I can enable KMonad with my laptop configuration files and it works as intended. The trackpad however, becomes unresponsive. Or rather, more precisely, pointer movements with the trackpad are impossible. The mouse cursor does not disappear from the screen and I still have the possibility of left- and right-clicking with the trackpad.

Just in case, I also ran these commands to give `uinput` perms to the mouse (because, despite its unresponsiveness, if I `sudo cat /dev/input/event4` or `sudo cat /dev/input/mouse0`, I could still see how my trackpad movements get intercepted):
```sh
sudo chgrp uinput /dev/input/by-path/platform-INT33C3:00-event-mouse && sudo chmod g+rw /dev/input/by-path/platform-INT33C3:00-event-mouse
sudo chgrp uinput /dev/input/mouse0 && sudo chmod g+rw /dev/input/mouse0
```
As you can guess, these commands did not help to restore trackpad usage.

At the moment, my only method for restoring the trackpad functionality is to restart the whole computer because killing off KMonad is not enough.

**Comment by slotThe** [comment]
On Tue, Oct 27 2020 03:31, precondition wrote:
> First of all, I did read the FAQ section on Linux permissions. I added my user to the `uinput` group. I also created a new file at `/lib/udev/rules.d/50-kmonad-rules.rules` and copy-pasted in the rules but that didn't seem to have an effect so I tried appending the rules  written in the FAQ to the `/lib/udev/rules.d/50-udev-default.rules` but that seems to have been in vain. I did also run `sudo modprobe uinput` but I didn't notice any effect from that.

Just to rule out obvious errors first, make sure that the uinput group
actually exists and you are properly added as a member (I know that you
said you read the FAQ, but you didn't mention that you created the
group; just making sure).  Running `groups your-user-name` should give
you that information).

Some distros only have an input group by default; in that case you don't
necessarily have to create the uinput group at all, simply change the
GROUP parameter to the correct one in the udev rule (again, make sure
you're added to the group etc.).

**Comment by precondition** [comment]
> On Tue, Oct 27 2020 03:31, precondition wrote: First of all, I did read the FAQ section on Linux permissions. I added my user to the `uinput` group. I also created a new file at `/lib/udev/rules.d/50-kmonad-rules.rules` and copy-pasted in the rules but that didn't seem to have an effect so I tried appending the rules written in the FAQ to the `/lib/udev/rules.d/50-udev-default.rules` but that seems to have been in vain. I did also run `sudo modprobe uinput` but I didn't notice any effect from that.
> Just to rule out obvious errors first, make sure that the uinput group actually exists and you are properly added as a member (I know that you said you read the FAQ, but you didn't mention that you created the group; just making sure). Running `groups your-user-name` should give you that information). Some distros only have an input group by default; in that case you don't necessarily have to create the uinput group at all, simply change the GROUP parameter to the correct one in the udev rule (again, make sure you're added to the group etc.).

There was no pre-existing `input` group so I created a new `uinput` group and added my user to it.

**Comment by slotThe** [comment]
What's the output of `sudo modprobe --first-time uinput`?  This should
fail if the uinput module is already loaded.

**Comment by precondition** [comment]
`modprobe: ERROR: could not insert 'uinput': Module already in kernel`

**Comment by Swoogan** [comment]
I have the same problem on Kubuntu. I added myself to the `uinput` group. I get this when I run the command without sudo:

> /dev/input/by-path/platform-i8042-serio-0-event-kbd: openFd: permission denied (Permission denied)

With sudo, the program works, but my terminal constantly scrolls and my trackpad stops working.

**Comment by slotThe** [comment]
On Thu, Jan 21 2021 17:54, Colin Svingen wrote:

> I have the same problem on Kubuntu. I added myself to the `uinput`
> group. I get this when I run the command without sudo:
>
>> /dev/input/by-path/platform-i8042-serio-0-event-kbd: openFd: permission denied (Permission denied)
>
> With sudo, the program works, but my terminal constantly scrolls and
> my trackpad stops working.

I presume the `uinput' group is actually what you want?  Do you have
something like 

    KERNEL=="uinput", GROUP="uinput"

somewhere within some rules file?

Does the above `modprobe' command return, that the module is loaded?

**Comment by Swoogan** [comment]
I have setup my system according to the [FAQ](https://github.com/david-janssen/kmonad/blob/master/doc/faq.md#q-how-do-i-get-uinput-permissions)

```
~ $ cat /etc/udev/rules.d/70-kmonad-uinput.rules 
KERNEL=="uinput", MODE="0660", GROUP="uinput", OPTIONS+="static_node=uinput"
```

```
~ $ sudo modprobe --first-time uinput
modprobe: ERROR: could not insert 'uinput': Module already in kernel
```

```
~ $ groups
adm cdrom sudo dip plugdev lpadmin sambashare uinput 
```

Let me know if there is anything else I can check.

**Comment by slotThe** [comment]
That is very weird indeed; perhaps this is actually the kernel changing under our feet (or, well, Ubuntu changing the kernel under our feet, as simply doing what's described in the FAQ works on all of the machines I've tested it on)

Can you try what @MuhammedZakir linked [here](https://github.com/david-janssen/kmonad/issues/160#issuecomment-766121884) and see if that helps?

@precondition if you are still having this issue, would you care to do the same?

**Comment by Swoogan** [comment]
I followed the instructions in the other issue and now I no longer need to `sudo` to start kmonad. However, the trackpad still becomes unresponsive. On the other hand, I have discovered that I can make responsive again by running:

    sudo rmmod uinput

which complains with the following, but still re-enables the trackpad:

> rmmod: ERROR: Module uinput is builtin.

**Comment by Swoogan** [comment]
I have good news! I cloned the repo to see if we could start to debug the problem and found that it's not present in the source code. I'm able to use the binary created with `stack` without any issues.

**Comment by sebastiansam55** [comment]
Hey so I'm not super familiar with how `kmonad` functions low level but I sort of ran across this issue like this earlier and thought I would throw my two cents in. 

If `kmonad` is creating a new uinput device that functions as a trackpad the signals that are sent are different. EG a mouse capabilities look like this:
```python
                    cap[e.EV_REL] = [
                        e.REL_X,
                        e.REL_Y,
                        e.REL_HWHEEL,
                        e.REL_WHEEL,
                        e.REL_HWHEEL_HI_RES,
                        e.REL_WHEEL_HI_RES
                    ]
```
and a trackpad looks like:
```python
                    cap[e.EV_ABS] = [
                        (e.ABS_X, AbsInfo(value=3800, min=1250, max=5631, fuzz=0, flat=0, resolution=59)),
                        (e.ABS_Y, AbsInfo(value=3500, min=1205, max=4834, fuzz=0, flat=0, resolution=81)),
                        (e.ABS_PRESSURE, AbsInfo(value=1, min=0, max=255, fuzz=0, flat=0, resolution=0)),
                        (e.ABS_TOOL_WIDTH, AbsInfo(value=0, min=0, max=15, fuzz=0, flat=0, resolution=0)),
                        (e.ABS_MT_SLOT, AbsInfo(value=0, min=0, max=1, fuzz=0, flat=0, resolution=0)),
                        (e.ABS_MT_POSITION_X, AbsInfo(value=0, min=1250, max=5631, fuzz=0, flat=0, resolution=59)),
                        (e.ABS_MT_POSITION_Y, AbsInfo(value=0, min=1205, max=4834, fuzz=0, flat=0, resolution=81)),
                        (e.ABS_MT_TRACKING_ID, AbsInfo(value=0, min=0, max=65535, fuzz=0, flat=0, resolution=0))
                    ]
```

**Comment by david-janssen** [comment]
Just popping in to say that the `uinput` device created by KMonad is only a keyboard. We don't do anything with mice or trackpads.

Additionally, as I think compiling from source has solved the issue, can I close this?

**Comment by precondition** [comment]
First, I've pulled the latest changes on master and compiled an up-to-date build using `stack build`.
Next, I've tried the steps outlined [here](https://github.com/david-janssen/kmonad/issues/160#issuecomment-766121884) which solved my permissions issues. I can now run kmonad without `chgrp`'ing and `chown`'ing root files manually beforehand.

However, my trackpad still gets disabled. Most of the time, when I *first* start KMonad, the trackpad works fine but if I restart KMonad, it always disables the trackpad.

Meanwhile, I found a way to re-enable the trackpad without rebooting the entire computer: `synclient TouchpadOff=0`. For some reasons, it seems like the `TouchpadOff` field gets ticked when I start up KMonad.

PS: Sorry for disappearing

**Comment by ceblan** [comment]
My five cents to this issue, I hope it helps:
I was having the same problem as described by @Swoogan  (terminal continuously scrolling and trackpad unusable) and also in kubuntu. I am not sure if the problem is KDE or Ubuntu messing with drivers, but I tried to launch kmonad in a non-X11 console (via Alt-F2) and the problem didn't happen!. Then, I set it up to be launched via systemd as per the manual and... voilà... kmonad is working and so is the trackpad.

**Comment by david-janssen** [comment]
> PS: Sorry for disappearing

Zero worries, I do the same.

Is this still an issue, are you still running into this?

**Comment by precondition** [comment]
> > PS: Sorry for disappearing
> 
> Zero worries, I do the same.
> 
> Is this still an issue, are you still running into this?

Yes I'm still running into the issue even after rebuilding a binary from latest master commits.

What's interesting is that I can run the program `./kmonad laptop.kbd`, notice that the touch pad ceased working. Consequently, run `synclient TouchpadOff=0` and then I can interrupt KMonad, restart it as many times as I want and the touchpad will keep working but if I do the slightest of changes to my `laptop.kbd` file, the touchpad fails to work again. The change doesn't even have to be meaningful. Simply removing comments is enough to break the touchpad.

**Comment by david-janssen** [comment]
We just had an issue that *might* be related. Maybe it has something to do with palm-detection and the OS turning of the touchpad when you type? Do you have this enabled in your X11?

https://github.com/kmonad/kmonad/issues/24#issuecomment-856090520

**Comment by precondition** [comment]
> We just had an issue that _might_ be related. Maybe it has something to do with palm-detection and the OS turning of the touchpad when you type? Do you have this enabled in your X11?
> 
> [#24 (comment)](https://github.com/kmonad/kmonad/issues/24#issuecomment-856090520)

Yes, I have the touchpad set to disable when typing.

**Comment by david-janssen** [comment]
Could you try disabling this for a moment to see if your issue disappears? Not as a fix, but as a diagnostic effort. If that *does* fix it, then we at least know what is causing the problem. Also, are you launching kmonad with a delay? I recommend either manually including a `sleep 1 && ...` section, or using the `-w` flag in newer builds. Maybe you aren't releasing all the keys on your hardware keyboard, causing the OS to register continuous keypresses, and your touchpad disabler to keep turning off your touchpad.

**Comment by precondition** [comment]
I can confirm that the issue stops when I *don't* disable the touchpad when typing.

PS: I'm closing because I take far too long to respond anyways

**Comment by SalmanFarooqShiekh** [comment]
I can confirm that the issue as described by @Swoogan and @ceblan i.e.:
> With sudo, the program works, but my terminal constantly scrolls and my trackpad stops working.
is fixed by adding a short sleep interval:
`sleep 1; kmonad your/own/config.kbd`
as described by @david-janssen 
> Could you try disabling this for a moment to see if your issue disappears? Not as a fix, but as a diagnostic effort. If that _does_ fix it, then we at least know what is causing the problem. Also, are you launching kmonad with a delay? I recommend either manually including a `sleep 1 && ...` section, or using the `-w` flag in newer builds. Maybe you aren't releasing all the keys on your hardware keyboard, causing the OS to register continuous keypresses, and your touchpad disabler to keep turning off your touchpad.

**Comment by david-janssen** [comment]
@SalmanFarooqShiekh 

Yep, this is such a common and inevitable problem that we're added a `start-delay` option, which can be set through the command-line invocation or in your `defcfg` block, IIRC.

**Comment by Swoogan** [comment]
I pulled the latest code and the issue seems to be resolved (I know, I've said this once before...). The default delay (300ms) seems to be enough time to prevent the issue that I didn't have to explicitly set a delay anywhere.
